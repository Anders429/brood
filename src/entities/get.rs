//! Extracts a component list from a batch of entities.
//!
//! This is defined for every entity that contains a given component. If an entity does not contain
//! a component, calling `get()` for that component will fail to compile.

use alloc::vec::Vec;

/// Type marker for the location of a component.
///
/// This does not actually have to be specified when calling `get()`. The compiler can infer its
/// location.
pub enum Index {}

/// Extracts a list of components from entities.
///
/// Removes the requested component, returning the remainder of the entities with the component
/// removed.
///
/// If a component is not contained in the entity, attempting to `get()` it with this trait will
/// fail to compile. This is because it is defined recursively on either the component itself or
/// lists that have the component within them.
///
/// This is generic over a component `T` and an index `I` (denoting the location of the component).
pub trait Get<T, I> {
    /// The entities returned after the component `T` is extracted.
    type Remainder;

    /// Remove the component list `Vec<T>` from the entity.
    ///
    /// Consumes the entities, returning the component and the entities with the component removed.
    fn get(self) -> (Vec<T>, Self::Remainder);
}

impl<E, T> Get<T, Index> for (Vec<T>, E) {
    type Remainder = E;

    fn get(self) -> (Vec<T>, Self::Remainder) {
        self
    }
}

impl<C, E, I, T> Get<T, (I,)> for (Vec<C>, E)
where
    E: Get<T, I>,
{
    type Remainder = (Vec<C>, <E as Get<T, I>>::Remainder);

    fn get(self) -> (Vec<T>, Self::Remainder) {
        let (target, remainder) = self.1.get();
        (target, (self.0, remainder))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::entities;
    use alloc::vec;

    #[derive(Clone, Debug, Eq, PartialEq)]
    struct A;
    #[derive(Clone, Debug, Eq, PartialEq)]
    struct B;
    #[derive(Clone, Debug, Eq, PartialEq)]
    struct C;

    #[test]
    fn get_a() {
        assert_eq!(
            entities!((A, B, C); 100).entities.get(),
            (vec![A; 100], entities!((B, C); 100).entities)
        );
    }

    #[test]
    fn get_b() {
        assert_eq!(
            entities!((A, B, C); 100).entities.get(),
            (vec![B; 100], entities!((A, C); 100).entities)
        );
    }

    #[test]
    fn get_c() {
        assert_eq!(
            entities!((A, B, C); 100).entities.get(),
            (vec![C; 100], entities!((A, B); 100).entities)
        );
    }
}
