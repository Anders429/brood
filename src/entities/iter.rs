use super::{raw, Entities};

pub struct Iter<E>
where
    E: raw::Entities,
{
    pub(crate) entities: E,
    len: usize,
    current_len: usize,
}

impl<E> Iter<E>
where
    E: raw::Entities,
{
    /// # Panics
    /// Panics if the columns are not all the same length.
    pub fn new<F>(entities: F) -> Self
    where
        F: Entities<Raw = E>,
    {
        assert!(entities.check_len());
        // SAFETY: We just guaranteed the lengths of all columns are equal.
        unsafe { Self::new_unchecked(entities) }
    }

    /// # Safety
    /// The caller must guarantee that the lengths of all columns within `entities` are equal.
    pub unsafe fn new_unchecked<F>(entities: F) -> Self
    where
        F: Entities<Raw = E>,
    {
        Self {
            len: entities.component_len(),
            current_len: entities.component_len(),
            // SAFETY: The component columns in `entities` all have the same length.
            entities: unsafe { entities.into_raw_entities_unchecked() },
        }
    }
}

impl<E> Iterator for Iter<E>
where
    E: raw::Entities,
{
    type Item = E::Entity;

    fn next(&mut self) -> Option<Self::Item> {
        if self.current_len > 0 {
            self.current_len -= 1;
            Some(
                // SAFETY: `self.entities` has exactly `self.current_len` rows left.
                unsafe { self.entities.next() },
            )
        } else {
            None
        }
    }
}

impl<E> Drop for Iter<E>
where
    E: raw::Entities,
{
    fn drop(&mut self) {
        // SAFETY: `self.entities` contains the raw parts for `Vec`s of length `self.len`.
        unsafe {
            self.entities.drop(self.len);
        }
    }
}
