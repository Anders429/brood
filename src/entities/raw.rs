use alloc::vec::Vec;

use crate::{component::Component, entity, hlist::define_null};
use core::{mem, mem::ManuallyDrop, ptr};

pub struct LengthlessIter<T> {
    start: *mut T,
    current: *mut T,
    capacity: usize,
}

impl<T> LengthlessIter<T> {
    pub(crate) fn from_vec(v: Vec<T>) -> Self {
        let mut v = ManuallyDrop::new(v);
        Self {
            start: v.as_mut_ptr(),
            current: v.as_mut_ptr(),
            capacity: v.capacity(),
        }
    }

    /// # Safety
    /// This `struct` must have been constructed from a `Vec` of length `len`.
    pub(crate) unsafe fn drop(&mut self, len: usize) {
        // SAFETY: `self.start` and `self.capacity`, together with the given `len`, form the raw
        // parts for a valid `Vec`.
        unsafe {
            Vec::from_raw_parts(self.start, len, self.capacity);
        }
    }
}

define_null!();

pub trait Entities {
    /// The entity type contained in this list of entities.
    type Entity: entity::Entity;

    /// # Safety
    /// The caller must guarantee that this is not called more times than there were elements in
    /// the original `Vec`s from which these iterators were created.
    unsafe fn next(&mut self) -> Self::Entity;

    /// # Safety
    /// `len` must be the length of the original `Vec`s from which these iterators were created.
    unsafe fn drop(&mut self, len: usize);
}

impl Entities for Null {
    type Entity = entity::Null;

    unsafe fn next(&mut self) -> Self::Entity {
        entity::Null
    }

    unsafe fn drop(&mut self, _len: usize) {}
}

impl<C, E> Entities for (LengthlessIter<C>, E)
where
    C: Component,
    E: Entities,
{
    type Entity = (C, E::Entity);

    unsafe fn next(&mut self) -> Self::Entity {
        (
            if mem::size_of::<C>() == 0 {
                // purposefully don't use 'current.offset' because for 0-size elements this would
                // return the same pointer.
                self.0.current = (self.0.current as *const i8).wrapping_offset(1) as *mut C;

                // Make up a value of this ZST.
                // SAFETY: Since this is a zero-sized type, this will always be valid.
                unsafe { mem::zeroed() }
            } else {
                let value = self.0.current;
                // SAFETY: By this method's safety contract, we know we are not at the end of this
                // allocation, and can therefore increment the pointer at least once more.
                self.0.current = unsafe { self.0.current.add(1) };
                // SAFETY: By this method's safety contract, we know that this value is within the
                // length of the original `Vec`, and can therefore be read validly.
                unsafe { ptr::read(value) }
            },
            // SAFETY: The safety invariants are upheld by this method's safety contract.
            unsafe { self.1.next() },
        )
    }

    unsafe fn drop(&mut self, len: usize) {
        // SAFETY: `len` is the correct length for this column's original `Vec`.
        unsafe {
            self.0.drop(len);
        }
        // SAFETY: The safety invariants are upheld by this method's safety contract.
        unsafe {
            self.1.drop(len);
        }
    }
}
