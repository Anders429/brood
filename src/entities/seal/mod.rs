mod into_raw;
mod length;
mod storage;

use crate::{component::Component, entities::Null};
use alloc::vec::Vec;
use into_raw::IntoRaw;
use length::Length;
use storage::Storage;

pub trait Seal: IntoRaw + Length + Storage {}

impl Seal for Null {}

impl<C, E> Seal for (Vec<C>, E)
where
    C: Component,
    E: Seal,
{
}
