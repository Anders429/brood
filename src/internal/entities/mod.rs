mod length;
mod storage;

use crate::{component::Component, entities::NullEntities};
use alloc::vec::Vec;
use length::EntitiesLength;
use storage::EntitiesStorage;

pub trait EntitiesSeal: EntitiesLength + EntitiesStorage {}

impl EntitiesSeal for NullEntities {}

impl<C, E> EntitiesSeal for (Vec<C>, E)
where
    C: Component,
    E: EntitiesSeal,
{
}
