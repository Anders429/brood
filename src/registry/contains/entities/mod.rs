mod sealed;

pub(crate) use sealed::Sealed;

/// Indicates that all of an entities' components are contained in the registry.
///
/// This allows reordering the components of the entities into the canonical ordering defined by
/// the registry.
///
/// If the entities contain components not in this registry, attempting to use this trait will
/// result in a compiler error, since the trait won't be implemented for the combination of entity
/// and registry.
pub trait ContainsEntities<Entities, Indices>: Sealed<Entities, Indices> {}

impl<Registry, Entities, Indices> ContainsEntities<Entities, Indices> for Registry where
    Registry: Sealed<Entities, Indices>
{
}
