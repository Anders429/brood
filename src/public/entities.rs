use crate::{component::Component, internal::entities::EntitiesSeal};
use alloc::vec::Vec;

#[derive(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct NullEntities;

#[cfg(feature = "serde")]
mod impl_serde {
    use crate::entities::NullEntities;
    use core::fmt;
    use serde::{de, de::Visitor, Deserialize, Deserializer, Serialize, Serializer};

    impl Serialize for NullEntities {
        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
        where
            S: Serializer,
        {
            serializer.serialize_unit_struct("NullEntities")
        }
    }

    impl<'de> Deserialize<'de> for NullEntities {
        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
        where
            D: Deserializer<'de>,
        {
            struct NullEntitiesVisitor;

            impl<'de> Visitor<'de> for NullEntitiesVisitor {
                type Value = NullEntities;

                fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                    formatter.write_str("struct NullEntities")
                }

                fn visit_unit<E>(self) -> Result<Self::Value, E>
                where
                    E: de::Error,
                {
                    Ok(NullEntities)
                }
            }

            deserializer.deserialize_unit_struct("NullEntities", NullEntitiesVisitor)
        }
    }
}

pub trait Entities: EntitiesSeal {}

impl Entities for NullEntities {}

impl<C, E> Entities for (Vec<C>, E)
where
    C: Component,
    E: Entities,
{
}

// TODO: Bikeshed this name. Yuck.
pub struct EntitiesIter<E>
where
    E: Entities,
{
    pub(crate) entities: E,
}

impl<E> EntitiesIter<E>
where
    E: Entities,
{
    pub fn new(entities: E) -> Self {
        assert!(entities.check_len());
        unsafe { Self::new_unchecked(entities) }
    }

    pub unsafe fn new_unchecked(entities: E) -> Self {
        Self { entities }
    }
}

#[macro_export]
macro_rules! entities {
    (($component:expr $(,$components:expr)* $(,)?); $n:expr) => {
        unsafe {
            $crate::entities::EntitiesIter::new_unchecked(
                ($crate::reexports::vec![$component; $n], entities!(@cloned ($($components),*); $n))
            )
        }
    };
    ($(($($components:expr),*)),* $(,)?) => {
        unsafe {
            $crate::entities::EntitiesIter::new_unchecked(
                entities!(@transpose [] $(($($components),*)),*)
            )
        }
    };
    ((); $n:expr) => {
        unsafe {
            $crate::entities::EntitiesIter::new_unchecked($crate::entities::NullEntities)
        }
    };
    () => {
        unsafe {
            $crate::entities::EntitiesIter::new_unchecked($crate::entities::NullEntities)
        }
    };
    (@cloned ($component:expr $(,$components:expr)* $(,)?); $n:expr) => {
        ($crate::reexports::vec![$component; $n], entities!(@cloned ($($components),*); $n))
    };
    (@cloned (); $n:expr) => {
        $crate::entities::NullEntities
    };
    (@transpose [$([$($column:expr),*])*] $(($component:expr $(,$components:expr)*  $(,)?)),*) => {
        entities!(@transpose [$([$($column),*])* [$($component),*]] $(($($components),*)),*)
    };
    (@transpose [$([$($column:expr),*])*] $(()),*) => {
        entities!(@as_vec ($(($($column),*)),*))
    };
    (@as_vec (($($column:expr),*) $(,($($columns:expr),*))* $(,)?)) => {
        ($crate::reexports::vec![$($column),*], entities!(@as_vec ($(($($columns),*)),*)))
    };
    (@as_vec ()) => {
        $crate::entities::NullEntities
    };
}
