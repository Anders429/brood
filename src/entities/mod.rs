//! A heterogeneous list of batches of [`Component`]s stored within a [`World`].
//!
//! A [`Batch`] of [`Entities`] is a container of entities all made of the same combination of
//! `Component`s. Inserting a `Batch` of `Entities` is much more efficient than storing them one at
//! a time.
//!
//! `Batch`es of entities are most often defined using the [`entities!`] macro. The items
//! contained within this module should rarely be needed in user code.
//!
//! `Entity`s are stored within [`World`]s to allow efficient querying and iteration with other
//! entities of similar components. Since entities are defined as heterogeneous lists, they can be
//! made of an arbitrary number of components. `World`s can store entities made up of any
//! combination of components, so long as those components are stored in the `World`'s
//! [`Registry`].
//!
//! # Example
//! ``` rust
//! use brood::{entities, registry, World};
//!
//! // Define components.
//! struct Foo(usize);
//! struct Bar(bool);
//!
//! type Registry = registry!(Foo, Bar);
//!
//! let mut world = World::<Registry>::new();
//!
//! world.extend(entities!((Foo(42), Bar(false)), (Foo(100), Bar(true))));
//! ```
//!
//! [`Batch`]: crate::entities::Batch
//! [`Component`]: crate::component::Component
//! [`Entities`]: crate::entities::Entities
//! [`entities!`]: crate::entities!
//! [`Registry`]: crate::registry::Registry
//! [`World`]: crate::world::World

mod seal;

use crate::{component::Component, hlist::define_null};
use alloc::vec::Vec;
use seal::Seal;

define_null!();

/// A heterogeneous list of columns of [`Component`]s.
///
/// In order for `Entities` to be able to be used, it must be contained within a [`Batch`]. This
/// guarnatees that the length of all columns within the heterogeneous list are equal.
///
/// Entities are stored within [`World`]s. In order for an entity to be able to be stored within a
/// `World`, that `World`'s [`Registry`] must include the `Component`s that make up an entity.
///
/// Note that entities must consist of unique component types. Duplicate components are not
/// supported. When multiple components of the same type are included in an entity, a `World` will
/// only store one of those components.
///
/// # Example
/// ``` rust
/// use brood::entities;
///
/// // Define components.
/// struct Foo(usize);
/// struct Bar(bool);
///
/// // Creates `Entities` wrapped in a `Batch` struct.
/// let entities = entities!((Foo(42), Bar(true)), (Foo(100), Bar(false)));
/// ```
///
/// [`Batch`]: crate::entities::Batch
/// [`Component`]: crate::component::Component
/// [`Registry`]: crate::registry::Registry
/// [`World`]: crate::world::World
pub trait Entities: Seal {}

impl Entities for Null {}

impl<C, E> Entities for (Vec<C>, E)
where
    C: Component,
    E: Entities,
{
}

/// A batch of entity columns of unified length.
///
/// This is a wrapper for an [`Entities`] heterogeneous list of columns of components, with the
/// guarantee that all columns are of the same length. In other words, this is a collection of
/// entities separated column-wise into their components.
///
/// A `Batch` is most often created using the [`entities!`] macro.
///
/// # Example
/// ``` rust
/// use brood::entities;
///
/// // Define components.
/// struct Foo(usize);
/// struct Bar(bool);
///
/// // This defines a `Batch` of entities, with guaranteed equal column length.
/// let entities = entities!((Foo(42), Bar(false)), (Foo(100), Bar(true)));
/// ```
///
/// [`Entities`]: crate::entities::Entities
/// [`entities!`]: crate::entities!
pub struct Batch<E>
where
    E: Entities,
{
    pub(crate) entities: E,
    len: usize,
}

impl<E> Batch<E>
where
    E: Entities,
{
    /// Creates a new `Batch`, wrapping the given [`Entities`] heterogeneous list.
    ///
    /// This method performs a run-time check to ensure the columns are all the same length.
    ///
    /// Note that this method can only be called using a raw heterogeneous list of entity columns.
    /// There is currently no macro that can make the call simpler. It is recommended to use the
    /// [`entities!`] macro instead of this method.
    ///
    /// # Example
    /// ``` rust
    /// use brood::{entities, entities::Batch};
    ///
    /// // Define components.
    /// struct Foo(usize);
    /// struct Bar(bool);
    ///
    /// let batch = Batch::new((vec![42; 10], (vec![true; 10], entities::Null)));
    /// ```
    ///
    /// [`Entities`]: crate::entities::Entities
    /// [`entities!]: crate::entities!
    ///
    /// # Panics
    /// Panics if the columns are not all the same length.
    pub fn new(entities: E) -> Self {
        assert!(entities.check_len());
        unsafe { Self::new_unchecked(entities) }
    }

    /// Creates a new `Batch`, wrapping the given [`Entities`] heterogeneous list, skipping any
    /// run-time checks for column length.
    ///
    /// Note that this method can only be called using a raw heterogeneous list of entity columns.
    /// There is currently no macro that can make the call simpler. It is recommended to use the
    /// [`entities!`] macro instead of this method.
    ///
    /// # Safety
    /// The caller must guarantee that the lengths of all columns within `entities` are equal.
    ///
    /// # Example
    /// ``` rust
    /// use brood::{entities, entities::Batch};
    ///
    /// // Define components.
    /// struct Foo(usize);
    /// struct Bar(bool);
    ///
    /// let batch = unsafe { Batch::new_unchecked((vec![42; 10], (vec![true; 10], entities::Null))) };
    /// ```
    /// [`Entities`]: crate::entities::Entities
    /// [`entities!]: crate::entities!
    pub unsafe fn new_unchecked(entities: E) -> Self {
        Self {
            len: entities.component_len(),
            entities,
        }
    }

    pub(crate) fn len(&self) -> usize {
        self.len
    }
}

/// Creates a batch of entities made from the same components.
///
/// This macro allows multiple entities to be defined for efficient storage within a [`World`]. The
/// syntax is similar to an array expression or the [`vec!`] macro. There are two forms of this
/// macro:
///
/// # Create entities given a list of tuples of components
///
/// ``` rust
/// use brood::entities;
///
/// // Define components `Foo` and `Bar`.
/// struct Foo(u16);
/// struct Bar(f32);
///
/// let my_entities = entities![(Foo(42), Bar(1.5)), (Foo(4), Bar(1.0))];
/// ```
///
/// Note that all tuples must be made from the same components in the same order.
///
/// # Create entities from a tuple of components and a size
///
/// ``` rust
/// use brood::entities;
///
/// // Define components `Foo` and `Bar`.
/// #[derive(Clone)]
/// struct Foo(u16);
///
/// #[derive(Clone)]
/// struct Bar(f32);
///
/// let my_entities = entities![(Foo(42), Bar(1.5)); 3];
/// ```
///
/// This syntax only supports entities made from components that implement [`Clone`], and the size
/// does not have to be a constant value.
///
/// This will call `clone()` to duplicate each component, so one should be careful with types that
/// have a nonstandard `Clone` implementation. For example, using `Rc` as a component value in this
/// context will create multiple entities with references to the same boxed value, not multiple
/// references to independently boxed values.
///
/// Using `0` as a size is allowed, and produces a container of no entities. This will still
/// evaluate all expressions passed as components, however, and immediately drop the resulting
/// values, so be mindful of side effects.
///
/// [`Clone`]: core::clone::Clone
/// [`World`]: crate::World
/// [`vec!`]: alloc::vec!
#[macro_export]
macro_rules! entities {
    (($component:expr $(,$components:expr)* $(,)?); $n:expr) => {
        unsafe {
            $crate::entities::Batch::new_unchecked(
                ($crate::reexports::vec![$component; $n], $crate::entities!(@cloned ($($components),*); $n))
            )
        }
    };
    ($(($($components:expr),*)),+ $(,)?) => {
        unsafe {
            $crate::entities::Batch::new_unchecked(
                $crate::entities!(@transpose [] $(($($components),*)),+)
            )
        }
    };
    ((); $n:expr) => {
        unsafe {
            $crate::entities::Batch::new_unchecked($crate::entities::Null)
        }
    };
    () => {
        unsafe {
            $crate::entities::Batch::new_unchecked($crate::entities::Null)
        }
    };

    (@cloned ($component:expr $(,$components:expr)* $(,)?); $n:expr) => {
        ($crate::reexports::vec![$component; $n], $crate::entities!(@cloned ($($components),*); $n))
    };
    (@cloned (); $n:expr) => {
        $crate::entities::Null
    };

    (@transpose [$([$($column:expr),*])*] $(($component:expr $(,$components:expr)*  $(,)?)),*) => {
        $crate::entities!(@transpose [$([$($column),*])* [$($component),*]] $(($($components),*)),*)
    };
    (@transpose [$([$($column:expr),*])*] $(()),*) => {
        $crate::entities!(@as_vec ($(($($column),*)),*))
    };

    (@as_vec (($($column:expr),*) $(,($($columns:expr),*))* $(,)?)) => {
        ($crate::reexports::vec![$($column),*], $crate::entities!(@as_vec ($(($($columns),*)),*)))
    };
    (@as_vec ()) => {
        $crate::entities::Null
    };
}
