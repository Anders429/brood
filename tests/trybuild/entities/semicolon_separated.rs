use brood::entities;

// Define components.
#[derive(Clone)]
struct A;
#[derive(Clone)]
struct B;

fn main () {
    let entities = entities!((A, B); (A, B); (A, B));
}
