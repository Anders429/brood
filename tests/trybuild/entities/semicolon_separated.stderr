error: no rules expected the token `;`
  --> tests/trybuild/entities/semicolon_separated.rs:10:44
   |
10 |     let entities = entities!((A, B); (A, B); (A, B));
   |                                            ^ no rules expected this token in macro call
